// ts-morph/dist/utils/TypeScriptVersionChecker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = require("../typescript");
const [tsMajor, tsMinor, tsPatch] = typescript_1.ts.version.split(".").map(v => parseInt(v, 10));
class TypeScriptVersionChecker {
    constructor() {
    }
    static isEqual(major, minor, patch) {
        return tsMajor === major && tsMinor === minor && tsPatch === patch;
    }
    static isLessThan(major, minor, patch) {
        if (tsMajor < major)
            return true;
        if (tsMajor === major && tsMinor < minor)
            return true;
        return tsMajor === major && tsMinor === minor && tsPatch < patch;
    }
    static isLessThanOrEqual(major, minor, patch) {
        return TypeScriptVersionChecker.isLessThan(major, minor, patch)
            || TypeScriptVersionChecker.isEqual(major, minor, patch);
    }
    static isGreaterThan(major, minor, patch) {
        return !TypeScriptVersionChecker.isLessThanOrEqual(major, minor, patch);
    }
    static isGreaterThanOrEqual(major, minor, patch) {
        return !TypeScriptVersionChecker.isLessThan(major, minor, patch);
    }
}
exports.TypeScriptVersionChecker = TypeScriptVersionChecker;
