// ts-morph/dist/compiler/ast/literal/template/TemplateExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const manipulation_1 = require("../../../../manipulation");
const expression_1 = require("../../expression");
exports.TemplateExpressionBase = expression_1.PrimaryExpression;
class TemplateExpression extends exports.TemplateExpressionBase {
    /**
     * Gets the template head.
     */
    getHead() {
        return this._getNodeFromCompilerNode(this.compilerNode.head);
    }
    /**
     * Gets the template spans.
     */
    getTemplateSpans() {
        return this.compilerNode.templateSpans.map(s => this._getNodeFromCompilerNode(s));
    }
    /**
     * Sets the literal value.
     *
     * Note: This could possibly replace the node if you remove all the tagged templates.
     * @param value - Value to set.
     * @returns The new node if the kind changed; the current node otherwise.
     */
    setLiteralValue(value) {
        const childIndex = this.getChildIndex();
        const parent = this.getParentSyntaxList() || this.getParentOrThrow();
        manipulation_1.replaceNodeText({
            sourceFile: this._sourceFile,
            start: this.getStart() + 1,
            replacingLength: this.getWidth() - 2,
            newText: value
        });
        return parent.getChildAtIndex(childIndex);
    }
}
exports.TemplateExpression = TemplateExpression;
