// ts-morph/dist/compiler/ast/literal/StringLiteral.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const manipulation_1 = require("../../../manipulation");
const utils_1 = require("../../../utils");
const expression_1 = require("../expression");
const QuoteKind_1 = require("./QuoteKind");
exports.StringLiteralBase = expression_1.LiteralExpression;
class StringLiteral extends exports.StringLiteralBase {
    /**
     * Gets the literal value.
     *
     * This is equivalent to .getLiteralText() for string literals and only exists for consistency with other literals.
     */
    getLiteralValue() {
        return this.compilerNode.text;
    }
    /**
     * Sets the literal value.
     * @param value - Value to set.
     */
    setLiteralValue(value) {
        manipulation_1.replaceNodeText({
            sourceFile: this._sourceFile,
            start: this.getStart() + 1,
            replacingLength: this.getWidth() - 2,
            newText: utils_1.StringUtils.escapeForWithinString(value, this.getQuoteKind())
        });
        return this;
    }
    /**
     * Gets the quote kind.
     */
    getQuoteKind() {
        return this.getText()[0] === "'" ? QuoteKind_1.QuoteKind.Single : QuoteKind_1.QuoteKind.Double;
    }
}
exports.StringLiteral = StringLiteral;
