// ts-morph/dist/utils/tsconfig/getTsParseConfigHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const compiler_1 = require("../compiler");
function getTsParseConfigHost(fileSystemWrapper, options) {
    const directories = [];
    const useCaseSensitiveFileNames = false; // shouldn't this be true? (it was false like this in the compiler)
    const host = {
        useCaseSensitiveFileNames,
        readDirectory: (rootDir, extensions, excludes, includes, depth) => {
            const result = compiler_1.readDirectory(fileSystemWrapper, useCaseSensitiveFileNames, rootDir, extensions, excludes, includes, depth);
            directories.push(...result.directories);
            return result.files;
        },
        fileExists: path => fileSystemWrapper.fileExistsSync(path),
        readFile: path => fileSystemWrapper.readFileSync(path, options.encoding),
        getDirectories: () => [...directories],
        clearDirectories: () => directories.length = 0
    };
    return host;
}
exports.getTsParseConfigHost = getTsParseConfigHost;
