// ts-morph/dist/compiler/ast/expression/object/ObjectLiteralExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors = require("../../../../errors");
const manipulation_1 = require("../../../../manipulation");
const structurePrinters_1 = require("../../../../structurePrinters");
const typescript_1 = require("../../../../typescript");
const utils_1 = require("../../../../utils");
const utils_2 = require("../../utils");
const PrimaryExpression_1 = require("../PrimaryExpression");
exports.ObjectLiteralExpressionBase = PrimaryExpression_1.PrimaryExpression;
class ObjectLiteralExpression extends exports.ObjectLiteralExpressionBase {
    getPropertyOrThrow(nameOrFindFunction) {
        return errors.throwIfNullOrUndefined(this.getProperty(nameOrFindFunction), () => utils_1.getNotFoundErrorMessageForNameOrFindFunction("property", nameOrFindFunction));
    }
    getProperty(nameOrFindFunction) {
        let findFunc;
        if (typeof nameOrFindFunction === "string")
            findFunc = prop => {
                if (prop["getName"] == null)
                    return false;
                return prop.getName() === nameOrFindFunction;
            };
        else
            findFunc = nameOrFindFunction;
        return this.getProperties().find(findFunc);
    }
    /**
     * Gets the properties.
     */
    getProperties() {
        return this.compilerNode.properties.map(p => this._getNodeFromCompilerNode(p));
    }
    /**
     * Gets the properties with comment object literal elements.
     */
    getPropertiesWithComments() {
        const members = utils_2.ExtendedParser.getContainerArray(this.compilerNode, this.getSourceFile().compilerNode);
        return members.map(p => this._getNodeFromCompilerNode(p));
    }
    /** @internal */
    _getAddIndex() {
        const members = utils_2.ExtendedParser.getContainerArray(this.compilerNode, this.getSourceFile().compilerNode);
        return members.length;
    }
    /**
     * Adds the specified property to the object literal expression.
     *
     * Note: If you only want to add a property assignment, then it might be more convenient to use #addPropertyAssignment.
     * @structure - The structure to add.
     */
    addProperty(structure) {
        return this.insertProperties(this._getAddIndex(), [structure])[0];
    }
    /**
     * Adds the specified properties to the object literal expression.
     *
     * Note: If you only want to add property assignments, then it might be more convenient to use #addPropertyAssignments.
     * @structures - The structures to add.
     */
    addProperties(structures) {
        return this.insertProperties(this._getAddIndex(), structures);
    }
    /**
     * Inserts the specified property to the object literal expression.
     *
     * Note: If you only want to insert a property assignment, then it might be more convenient to use #insertPropertyAssignment.
     * @index - The index to insert at.
     * @structure - The structure to insert.
     */
    insertProperty(index, structure) {
        return this.insertProperties(index, [structure])[0];
    }
    /**
     * Inserts the specified properties to the object literal expression.
     *
     * Note: If you only want to insert property assignments, then it might be more convenient to use #insertPropertyAssignments.
     * @index - The index to insert at.
     * @structures - The structures to insert.
     */
    insertProperties(index, structures) {
        const properties = this.getPropertiesWithComments();
        index = manipulation_1.verifyAndGetIndex(index, properties.length);
        // create property code
        const writer = this._getWriterWithChildIndentation();
        const structurePrinter = this._context.structurePrinterFactory.forObjectLiteralExpressionProperty();
        structurePrinter.printTexts(writer, structures);
        // insert
        manipulation_1.insertIntoCommaSeparatedNodes({
            parent: this.getChildSyntaxListOrThrow(),
            currentNodes: properties,
            insertIndex: index,
            newText: writer.toString(),
            useNewLines: true
        });
        // get the properties
        return manipulation_1.getNodesToReturn(properties, this.getPropertiesWithComments(), index, true);
    }
    /* Property Assignments */
    /**
     * Adds a property assignment.
     * @param structure - Structure that represents the property assignment to add.
     */
    addPropertyAssignment(structure) {
        return this.addPropertyAssignments([structure])[0];
    }
    /**
     * Adds property assignments.
     * @param structures - Structure that represents the property assignments to add.
     */
    addPropertyAssignments(structures) {
        return this.insertPropertyAssignments(this._getAddIndex(), structures);
    }
    /**
     * Inserts a property assignment at the specified index.
     * @param index - Child index to insert at.
     * @param structure - Structure that represents the property assignment to insert.
     */
    insertPropertyAssignment(index, structure) {
        return this.insertPropertyAssignments(index, [structure])[0];
    }
    /**
     * Inserts property assignments at the specified index.
     * @param index - Child index to insert at.
     * @param structures - Structures that represent the property assignments to insert.
     */
    insertPropertyAssignments(index, structures) {
        return this._insertProperty(index, structures, () => this._context.structurePrinterFactory.forPropertyAssignment());
    }
    /* Shorthand Property Assignments */
    /**
     * Adds a shorthand property assignment.
     * @param structure - Structure that represents the shorthand property assignment to add.
     */
    addShorthandPropertyAssignment(structure) {
        return this.addShorthandPropertyAssignments([structure])[0];
    }
    /**
     * Adds shorthand property assignments.
     * @param structures - Structure that represents the shorthand property assignments to add.
     */
    addShorthandPropertyAssignments(structures) {
        return this.insertShorthandPropertyAssignments(this._getAddIndex(), structures);
    }
    /**
     * Inserts a shorthand property assignment at the specified index.
     * @param index - Child index to insert at.
     * @param structure - Structure that represents the shorthand property assignment to insert.
     */
    insertShorthandPropertyAssignment(index, structure) {
        return this.insertShorthandPropertyAssignments(index, [structure])[0];
    }
    /**
     * Inserts shorthand property assignments at the specified index.
     * @param index - Child index to insert at.
     * @param structures - Structures that represent the shorthand property assignments to insert.
     */
    insertShorthandPropertyAssignments(index, structures) {
        return this._insertProperty(index, structures, () => this._context.structurePrinterFactory.forShorthandPropertyAssignment());
    }
    /* Spread Assignments */
    /**
     * Adds a spread assignment.
     * @param structure - Structure that represents the spread assignment to add.
     */
    addSpreadAssignment(structure) {
        return this.addSpreadAssignments([structure])[0];
    }
    /**
     * Adds spread assignments.
     * @param structures - Structure that represents the spread assignments to add.
     */
    addSpreadAssignments(structures) {
        return this.insertSpreadAssignments(this._getAddIndex(), structures);
    }
    /**
     * Inserts a spread assignment at the specified index.
     * @param index - Child index to insert at.
     * @param structure - Structure that represents the spread assignment to insert.
     */
    insertSpreadAssignment(index, structure) {
        return this.insertSpreadAssignments(index, [structure])[0];
    }
    /**
     * Inserts spread assignments at the specified index.
     * @param index - Child index to insert at.
     * @param structures - Structures that represent the spread assignments to insert.
     */
    insertSpreadAssignments(index, structures) {
        return this._insertProperty(index, structures, () => this._context.structurePrinterFactory.forSpreadAssignment());
    }
    /* Method Declarations */
    /**
     * Adds a method.
     * @param structure - Structure that represents the method to add.
     */
    addMethod(structure) {
        return this.addMethods([structure])[0];
    }
    /**
     * Adds methods.
     * @param structures - Structure that represents the methods to add.
     */
    addMethods(structures) {
        return this.insertMethods(this._getAddIndex(), structures);
    }
    /**
     * Inserts a method at the specified index.
     * @param index - Child index to insert at.
     * @param structure - Structure that represents the method to insert.
     */
    insertMethod(index, structure) {
        return this.insertMethods(index, [structure])[0];
    }
    /**
     * Inserts methods at the specified index.
     * @param index - Child index to insert at.
     * @param structures - Structures that represent the methods to insert.
     */
    insertMethods(index, structures) {
        return this._insertProperty(index, structures, () => this._context.structurePrinterFactory.forMethodDeclaration({ isAmbient: false }));
    }
    /* Get Accessor Declarations */
    /**
     * Adds a get accessor.
     * @param structure - Structure that represents the property assignment to add.
     */
    addGetAccessor(structure) {
        return this.addGetAccessors([structure])[0];
    }
    /**
     * Adds get accessors.
     * @param structures - Structure that represents the get accessors to add.
     */
    addGetAccessors(structures) {
        return this.insertGetAccessors(this._getAddIndex(), structures);
    }
    /**
     * Inserts a get accessor at the specified index.
     * @param index - Child index to insert at.
     * @param structure - Structure that represents the get accessor to insert.
     */
    insertGetAccessor(index, structure) {
        return this.insertGetAccessors(index, [structure])[0];
    }
    /**
     * Inserts get accessors at the specified index.
     * @param index - Child index to insert at.
     * @param structures - Structures that represent the get accessors to insert.
     */
    insertGetAccessors(index, structures) {
        return this._insertProperty(index, structures, () => this._context.structurePrinterFactory.forGetAccessorDeclaration({ isAmbient: false }));
    }
    /* Set Accessor Declarations */
    /**
     * Adds a set accessor.
     * @param structure - Structure that represents the property assignment to add.
     */
    addSetAccessor(structure) {
        return this.addSetAccessors([structure])[0];
    }
    /**
     * Adds set accessors.
     * @param structures - Structure that represents the set accessors to add.
     */
    addSetAccessors(structures) {
        return this.insertSetAccessors(this._getAddIndex(), structures);
    }
    /**
     * Inserts a set accessor at the specified index.
     * @param index - Child index to insert at.
     * @param structure - Structure that represents the set accessor to insert.
     */
    insertSetAccessor(index, structure) {
        return this.insertSetAccessors(index, [structure])[0];
    }
    /**
     * Inserts set accessors at the specified index.
     * @param index - Child index to insert at.
     * @param structures - Structures that represent the set accessors to insert.
     */
    insertSetAccessors(index, structures) {
        return this._insertProperty(index, structures, () => this._context.structurePrinterFactory.forSetAccessorDeclaration({ isAmbient: false }));
    }
    /**
     * @internal
     */
    _insertProperty(index, structures, createStructurePrinter) {
        index = manipulation_1.verifyAndGetIndex(index, this._getAddIndex());
        const writer = this._getWriterWithChildIndentation();
        const structurePrinter = new structurePrinters_1.CommaNewLineSeparatedStructuresPrinter(createStructurePrinter());
        const oldProperties = this.getPropertiesWithComments();
        structurePrinter.printText(writer, structures);
        manipulation_1.insertIntoCommaSeparatedNodes({
            parent: this.getFirstChildByKindOrThrow(typescript_1.SyntaxKind.SyntaxList),
            currentNodes: oldProperties,
            insertIndex: index,
            newText: writer.toString(),
            useNewLines: true
        });
        return manipulation_1.getNodesToReturn(oldProperties, this.getPropertiesWithComments(), index, false);
    }
}
exports.ObjectLiteralExpression = ObjectLiteralExpression;
