// ts-morph/dist/compiler/ast/base/name/NameableNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors = require("../../../../errors");
const manipulation_1 = require("../../../../manipulation");
const typescript_1 = require("../../../../typescript");
const utils_1 = require("../../../../utils");
const callBaseSet_1 = require("../../callBaseSet");
const ReferenceFindableNode_1 = require("./ReferenceFindableNode");
const callBaseGetStructure_1 = require("../../callBaseGetStructure");
const RenameableNode_1 = require("./RenameableNode");
function NameableNode(Base) {
    return NameableNodeInternal(ReferenceFindableNode_1.ReferenceFindableNode(RenameableNode_1.RenameableNode(Base)));
}
exports.NameableNode = NameableNode;
function NameableNodeInternal(Base) {
    return class extends Base {
        getNameNode() {
            return this._getNodeFromCompilerNodeIfExists(this.compilerNode.name);
        }
        getNameNodeOrThrow() {
            return errors.throwIfNullOrUndefined(this.getNameNode(), "Expected to have a name node.");
        }
        getName() {
            const identifier = this.getNameNode();
            return identifier == null ? undefined : identifier.getText();
        }
        getNameOrThrow() {
            return errors.throwIfNullOrUndefined(this.getName(), "Expected to have a name.");
        }
        rename(newName) {
            if (newName === this.getName())
                return this;
            if (utils_1.StringUtils.isNullOrWhitespace(newName)) {
                this.removeName();
                return this;
            }
            const nameNode = this.getNameNode();
            if (nameNode == null)
                addNameNode(this, newName);
            else
                Base.prototype.rename.call(this, newName);
            return this;
        }
        removeName() {
            const nameNode = this.getNameNode();
            if (nameNode == null)
                return this;
            manipulation_1.removeChildren({ children: [nameNode], removePrecedingSpaces: true });
            return this;
        }
        set(structure) {
            callBaseSet_1.callBaseSet(Base.prototype, this, structure);
            if (structure.name != null) {
                errors.throwIfWhitespaceOrNotString(structure.name, "structure.name");
                const nameNode = this.getNameNode();
                if (nameNode == null)
                    addNameNode(this, structure.name);
                else
                    nameNode.replaceWithText(structure.name);
            }
            else if (structure.hasOwnProperty("name"))
                this.removeName();
            return this;
        }
        getStructure() {
            return callBaseGetStructure_1.callBaseGetStructure(Base.prototype, this, {
                name: this.getName()
            });
        }
    };
}
function addNameNode(node, newName) {
    const openParenToken = node.getFirstChildByKindOrThrow(typescript_1.SyntaxKind.OpenParenToken);
    manipulation_1.insertIntoParentTextRange({
        insertPos: openParenToken.getStart(),
        newText: " " + newName,
        parent: node
    });
}
