// ts-morph/dist/compiler/tools/results/DiagnosticMessageChain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Diagnostic message chain.
 */
class DiagnosticMessageChain {
    /** @private */
    constructor(compilerObject) {
        this._compilerObject = compilerObject;
    }
    /**
     * Gets the underlying compiler object.
     */
    get compilerObject() {
        return this._compilerObject;
    }
    /**
     * Gets the message text.
     */
    getMessageText() {
        return this.compilerObject.messageText;
    }
    /**
     * Gets th enext diagnostic message chain in the chain.
     */
    getNext() {
        const next = this.compilerObject.next;
        if (next == null)
            return undefined;
        return new DiagnosticMessageChain(next);
    }
    /**
     * Gets the code of the diagnostic message chain.
     */
    getCode() {
        return this.compilerObject.code;
    }
    /**
     * Gets the category of the diagnostic message chain.
     */
    getCategory() {
        return this.compilerObject.category;
    }
}
exports.DiagnosticMessageChain = DiagnosticMessageChain;
