// ts-morph/dist/options/ManipulationSettingsContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const compiler_1 = require("../compiler");
const errors = require("../errors");
const typescript_1 = require("../typescript");
const utils_1 = require("../utils");
const SettingsContainer_1 = require("./SettingsContainer");
/** Kinds of indentation */
var IndentationText;
(function (IndentationText) {
    /** Two spaces */
    IndentationText["TwoSpaces"] = "  ";
    /** Four spaces */
    IndentationText["FourSpaces"] = "    ";
    /** Eight spaces */
    IndentationText["EightSpaces"] = "        ";
    /** Tab */
    IndentationText["Tab"] = "\t";
})(IndentationText = exports.IndentationText || (exports.IndentationText = {}));
/**
 * Holds the manipulation settings.
 */
class ManipulationSettingsContainer extends SettingsContainer_1.SettingsContainer {
    constructor() {
        super({
            indentationText: IndentationText.FourSpaces,
            newLineKind: typescript_1.NewLineKind.LineFeed,
            quoteKind: compiler_1.QuoteKind.Double,
            insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces: true,
            usePrefixAndSuffixTextForRename: false
        });
    }
    /**
     * Gets the editor settings based on the current manipulation settings.
     */
    getEditorSettings() {
        if (this._editorSettings == null) {
            this._editorSettings = {};
            utils_1.fillDefaultEditorSettings(this._editorSettings, this);
        }
        return Object.assign({}, this._editorSettings);
    }
    /**
     * Gets the format code settings.
     */
    getFormatCodeSettings() {
        if (this._formatCodeSettings == null) {
            this._formatCodeSettings = Object.assign({}, this.getEditorSettings(), { insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces: this._settings.insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces });
        }
        return Object.assign({}, this._formatCodeSettings);
    }
    /**
     * Gets the user preferences.
     */
    getUserPreferences() {
        if (this._userPreferences == null) {
            this._userPreferences = {
                quotePreference: this.getQuoteKind() === compiler_1.QuoteKind.Double ? "double" : "single",
                providePrefixAndSuffixTextForRename: this.getUsePrefixAndSuffixTextForRename()
            };
        }
        return Object.assign({}, this._userPreferences);
    }
    /**
     * Gets the quote kind used for string literals.
     */
    getQuoteKind() {
        return this._settings.quoteKind;
    }
    /**
     * Gets the new line kind.
     */
    getNewLineKind() {
        return this._settings.newLineKind;
    }
    /**
     * Gets the new line kind as a string.
     */
    getNewLineKindAsString() {
        return utils_1.newLineKindToString(this.getNewLineKind());
    }
    /**
     * Gets the indentation text.
     */
    getIndentationText() {
        return this._settings.indentationText;
    }
    /**
     * Gets whether to use prefix and suffix text when renaming.
     */
    getUsePrefixAndSuffixTextForRename() {
        return this._settings.usePrefixAndSuffixTextForRename;
    }
    /**
     * Sets one or all of the settings.
     * @param settings - Settings to set.
     */
    set(settings) {
        super.set(settings);
        this._editorSettings = undefined;
        this._formatCodeSettings = undefined;
        this._userPreferences = undefined;
    }
    /**
     * @internal
     * Gets the indent size as represented in spaces.
     */
    _getIndentSizeInSpaces() {
        const indentationText = this.getIndentationText();
        switch (indentationText) {
            case IndentationText.EightSpaces:
                return 8;
            case IndentationText.FourSpaces:
                return 4;
            case IndentationText.TwoSpaces:
                return 2;
            case IndentationText.Tab:
                return 4; // most common
            default:
                return errors.throwNotImplementedForNeverValueError(indentationText);
        }
    }
}
exports.ManipulationSettingsContainer = ManipulationSettingsContainer;
