// ts-morph/dist/compiler/ast/class/SetAccessorDeclaration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors = require("../../../errors");
const structures_1 = require("../../../structures");
const typescript_1 = require("../../../typescript");
const base_1 = require("../base");
const callBaseSet_1 = require("../callBaseSet");
const function_1 = require("../function");
const base_2 = require("./base");
const callBaseGetStructure_1 = require("../callBaseGetStructure");
const ClassElement_1 = require("./ClassElement");
exports.SetAccessorDeclarationBase = base_1.ChildOrderableNode(base_1.TextInsertableNode(base_1.DecoratableNode(base_2.AbstractableNode(base_1.ScopedNode(base_1.StaticableNode(function_1.FunctionLikeDeclaration(base_1.BodyableNode(base_1.PropertyNamedNode(ClassElement_1.ClassElement)))))))));
class SetAccessorDeclaration extends exports.SetAccessorDeclarationBase {
    /**
     * Sets the node from a structure.
     * @param structure - Structure to set the node with.
     */
    set(structure) {
        callBaseSet_1.callBaseSet(exports.SetAccessorDeclarationBase.prototype, this, structure);
        return this;
    }
    /**
     * Gets the corresponding get accessor if one exists.
     */
    getGetAccessor() {
        const parent = this.getParentIfKindOrThrow(typescript_1.SyntaxKind.ClassDeclaration);
        const thisName = this.getName();
        return parent.getInstanceProperties().find(p => p.getKind() === typescript_1.SyntaxKind.GetAccessor && p.getName() === thisName);
    }
    /**
     * Gets the corresponding get accessor or throws if not exists.
     */
    getGetAccessorOrThrow() {
        return errors.throwIfNullOrUndefined(this.getGetAccessor(), () => `Expected to find a corresponding get accessor for ${this.getName()}.`);
    }
    /**
     * Gets the structure equivalent to this node.
     */
    getStructure() {
        return callBaseGetStructure_1.callBaseGetStructure(exports.SetAccessorDeclarationBase.prototype, this, {
            kind: structures_1.StructureKind.SetAccessor
        });
    }
}
exports.SetAccessorDeclaration = SetAccessorDeclaration;
