// ts-morph/dist/compiler/ast/binding/BindingElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors = require("../../../errors");
const base_1 = require("../base");
const common_1 = require("../common");
exports.BindingElementBase = base_1.InitializerExpressionableNode(base_1.BindingNamedNode(common_1.Node));
class BindingElement extends exports.BindingElementBase {
    /**
     * Gets the binding element's dot dot dot token (...) if it exists or throws if not.
     */
    getDotDotDotTokenOrThrow() {
        return errors.throwIfNullOrUndefined(this.getDotDotDotToken(), "Expected to find a dot dot dot token (...).");
    }
    /**
     * Gets the binding element's dot dot dot token (...) if it exists or returns undefined.
     */
    getDotDotDotToken() {
        return this._getNodeFromCompilerNodeIfExists(this.compilerNode.dotDotDotToken);
    }
    /**
     * Gets binding element's property name node or throws if not found.
     *
     * For example in `const { a: b } = { a: 5 }`, `a` would be the property name.
     */
    getPropertyNameNodeOrThrow() {
        return errors.throwIfNullOrUndefined(this.getPropertyNameNode(), "Expected to find a property name node.");
    }
    /**
     * Gets binding element's property name node or returns undefined if not found.
     *
     * For example in `const { a: b } = { a: 5 }`, `a` would be the property name.
     */
    getPropertyNameNode() {
        return this._getNodeFromCompilerNodeIfExists(this.compilerNode.propertyName);
    }
}
exports.BindingElement = BindingElement;
