// ts-morph/dist/structures/StructureKind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StructureKind;
(function (StructureKind) {
    StructureKind[StructureKind["CallSignature"] = 0] = "CallSignature";
    StructureKind[StructureKind["Class"] = 1] = "Class";
    StructureKind[StructureKind["ConstructSignature"] = 2] = "ConstructSignature";
    StructureKind[StructureKind["Constructor"] = 3] = "Constructor";
    StructureKind[StructureKind["ConstructorOverload"] = 4] = "ConstructorOverload";
    StructureKind[StructureKind["Decorator"] = 5] = "Decorator";
    StructureKind[StructureKind["Enum"] = 6] = "Enum";
    StructureKind[StructureKind["EnumMember"] = 7] = "EnumMember";
    StructureKind[StructureKind["ExportAssignment"] = 8] = "ExportAssignment";
    StructureKind[StructureKind["ExportDeclaration"] = 9] = "ExportDeclaration";
    StructureKind[StructureKind["ExportSpecifier"] = 10] = "ExportSpecifier";
    StructureKind[StructureKind["Function"] = 11] = "Function";
    StructureKind[StructureKind["FunctionOverload"] = 12] = "FunctionOverload";
    StructureKind[StructureKind["GetAccessor"] = 13] = "GetAccessor";
    StructureKind[StructureKind["ImportDeclaration"] = 14] = "ImportDeclaration";
    StructureKind[StructureKind["ImportSpecifier"] = 15] = "ImportSpecifier";
    StructureKind[StructureKind["IndexSignature"] = 16] = "IndexSignature";
    StructureKind[StructureKind["Interface"] = 17] = "Interface";
    StructureKind[StructureKind["JsxAttribute"] = 18] = "JsxAttribute";
    StructureKind[StructureKind["JsxSpreadAttribute"] = 19] = "JsxSpreadAttribute";
    StructureKind[StructureKind["JsxElement"] = 20] = "JsxElement";
    StructureKind[StructureKind["JsxSelfClosingElement"] = 21] = "JsxSelfClosingElement";
    StructureKind[StructureKind["JSDoc"] = 22] = "JSDoc";
    StructureKind[StructureKind["Method"] = 23] = "Method";
    StructureKind[StructureKind["MethodOverload"] = 24] = "MethodOverload";
    StructureKind[StructureKind["MethodSignature"] = 25] = "MethodSignature";
    StructureKind[StructureKind["Namespace"] = 26] = "Namespace";
    StructureKind[StructureKind["Parameter"] = 27] = "Parameter";
    StructureKind[StructureKind["Property"] = 28] = "Property";
    StructureKind[StructureKind["PropertyAssignment"] = 29] = "PropertyAssignment";
    StructureKind[StructureKind["PropertySignature"] = 30] = "PropertySignature";
    StructureKind[StructureKind["SetAccessor"] = 31] = "SetAccessor";
    StructureKind[StructureKind["ShorthandPropertyAssignment"] = 32] = "ShorthandPropertyAssignment";
    StructureKind[StructureKind["SourceFile"] = 33] = "SourceFile";
    StructureKind[StructureKind["SpreadAssignment"] = 34] = "SpreadAssignment";
    StructureKind[StructureKind["TypeAlias"] = 35] = "TypeAlias";
    StructureKind[StructureKind["TypeParameter"] = 36] = "TypeParameter";
    StructureKind[StructureKind["VariableDeclaration"] = 37] = "VariableDeclaration";
    StructureKind[StructureKind["VariableStatement"] = 38] = "VariableStatement";
})(StructureKind = exports.StructureKind || (exports.StructureKind = {}));
