// ts-morph/dist/compiler/ast/expression/ConditionalExpression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Expression_1 = require("./Expression");
exports.ConditionalExpressionBase = Expression_1.Expression;
class ConditionalExpression extends exports.ConditionalExpressionBase {
    /**
     * Gets the condition of the conditional expression.
     */
    getCondition() {
        return this._getNodeFromCompilerNode(this.compilerNode.condition);
    }
    /**
     * Gets the question token of the conditional expression.
     */
    getQuestionToken() {
        return this._getNodeFromCompilerNode(this.compilerNode.questionToken);
    }
    /**
     * Gets the when true expression of the conditional expression.
     */
    getWhenTrue() {
        return this._getNodeFromCompilerNode(this.compilerNode.whenTrue);
    }
    /**
     * Gets the colon token of the conditional expression.
     */
    getColonToken() {
        return this._getNodeFromCompilerNode(this.compilerNode.colonToken);
    }
    /**
     * Gets the when false expression of the conditional expression.
     */
    getWhenFalse() {
        return this._getNodeFromCompilerNode(this.compilerNode.whenFalse);
    }
}
exports.ConditionalExpression = ConditionalExpression;
