// ts-morph/dist/compiler/ast/statement/SwitchStatement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Statement_1 = require("./Statement");
class SwitchStatement extends Statement_1.Statement {
    /**
     * Gets this switch statement's expression.
     */
    getExpression() {
        return this._getNodeFromCompilerNode(this.compilerNode.expression);
    }
    /**
     * Gets this switch statement's case block.
     */
    getCaseBlock() {
        return this._getNodeFromCompilerNode(this.compilerNode.caseBlock);
    }
    /**
     * Gets the switch statement's case block's clauses.
     */
    getClauses() {
        // convenience method
        return this.getCaseBlock().getClauses();
    }
    /**
     * Removes the specified clause based on the provided index.
     * @param index - Index.
     */
    removeClause(index) {
        return this.getCaseBlock().removeClause(index);
    }
    /**
     * Removes the specified clauses based on the provided index range.
     * @param indexRange - Index range.
     */
    removeClauses(indexRange) {
        return this.getCaseBlock().removeClauses(indexRange);
    }
}
exports.SwitchStatement = SwitchStatement;
