// ts-morph/dist/factories/CompilerFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const compiler_1 = require("../compiler");
const utils_1 = require("../compiler/ast/utils");
const errors = require("../errors");
const typescript_1 = require("../typescript");
const manipulation_1 = require("../manipulation");
const utils_2 = require("../utils");
const DirectoryCache_1 = require("./DirectoryCache");
const ForgetfulNodeCache_1 = require("./ForgetfulNodeCache");
const kindToWrapperMappings_1 = require("./kindToWrapperMappings");
const DocumentRegistry_1 = require("./DocumentRegistry");
/**
 * Factory for creating compiler wrappers.
 * @internal
 */
class CompilerFactory {
    /**
     * Initializes a new instance of CompilerFactory.
     * @param context - Project context.
     */
    constructor(context) {
        this.context = context;
        this.sourceFileCacheByFilePath = new utils_2.KeyValueCache();
        this.diagnosticCache = new utils_2.WeakCache();
        this.definitionInfoCache = new utils_2.WeakCache();
        this.documentSpanCache = new utils_2.WeakCache();
        this.diagnosticMessageChainCache = new utils_2.WeakCache();
        this.jsDocTagInfoCache = new utils_2.WeakCache();
        this.signatureCache = new utils_2.WeakCache();
        this.symbolCache = new utils_2.WeakCache();
        this.symbolDisplayPartCache = new utils_2.WeakCache();
        this.referenceEntryCache = new utils_2.WeakCache();
        this.referencedSymbolCache = new utils_2.WeakCache();
        this.referencedSymbolDefinitionInfoCache = new utils_2.WeakCache();
        this.typeCache = new utils_2.WeakCache();
        this.typeParameterCache = new utils_2.WeakCache();
        this.nodeCache = new ForgetfulNodeCache_1.ForgetfulNodeCache();
        this.sourceFileAddedEventContainer = new utils_2.EventContainer();
        this.sourceFileRemovedEventContainer = new utils_2.EventContainer();
        this.documentRegistry = new DocumentRegistry_1.DocumentRegistry(context.fileSystemWrapper);
        this.directoryCache = new DirectoryCache_1.DirectoryCache(context);
        // prevent memory leaks when the document registry key changes by just reseting it
        this.context.compilerOptions._onModified(() => {
            // repopulate the cache
            const currentSourceFiles = this.sourceFileCacheByFilePath.getValuesAsArray();
            for (const sourceFile of currentSourceFiles) {
                // reparse the source files in the new document registry, then populate the cache with the new nodes
                manipulation_1.replaceSourceFileForCacheUpdate(sourceFile);
            }
        });
    }
    /**
     * Gets all the source files sorted by their directory depth.
     */
    *getSourceFilesByDirectoryDepth() {
        for (const dir of this.getDirectoriesByDepth())
            yield* dir.getSourceFiles();
    }
    /**
     * Gets the source file paths from the internal cache.
     */
    getSourceFilePaths() {
        return Array.from(this.sourceFileCacheByFilePath.getKeys());
    }
    /**
     * Gets the child directories of a directory.
     * @param dirPath - Directory path.
     */
    getChildDirectoriesOfDirectory(dirPath) {
        return this.directoryCache.getChildDirectoriesOfDirectory(dirPath);
    }
    /**
     * Gets the child source files of a directory.
     * @param dirPath - Directory path.
     */
    getChildSourceFilesOfDirectory(dirPath) {
        return this.directoryCache.getChildSourceFilesOfDirectory(dirPath);
    }
    /**
     * Occurs when a source file is added to the cache.
     * @param subscription - Subscripton.
     * @param subscribe - Whether to subscribe or unsubscribe (default to true).
     */
    onSourceFileAdded(subscription, subscribe = true) {
        if (subscribe)
            this.sourceFileAddedEventContainer.subscribe(subscription);
        else
            this.sourceFileAddedEventContainer.unsubscribe(subscription);
    }
    /**
     * Occurs when a source file is removed from the cache.
     * @param subscription - Subscripton.
     */
    onSourceFileRemoved(subscription) {
        this.sourceFileRemovedEventContainer.subscribe(subscription);
    }
    /**
     * Adds a source file by structure or text.
     * @param filePath - File path.
     * @param structureOrText - Structure or text.
     * @param options - Options.
     */
    createSourceFile(filePath, sourceFileText, options) {
        sourceFileText = sourceFileText instanceof Function ? utils_2.getTextFromStringOrWriter(this.context.createWriter(), sourceFileText) : sourceFileText || "";
        if (typeof sourceFileText === "string")
            return this.createSourceFileFromText(filePath, sourceFileText, options);
        const writer = this.context.createWriter();
        const structurePrinter = this.context.structurePrinterFactory.forSourceFile({
            isAmbient: utils_2.FileUtils.getExtension(filePath) === ".d.ts"
        });
        structurePrinter.printText(writer, sourceFileText);
        return this.createSourceFileFromText(filePath, writer.toString(), options);
    }
    /**
     * Creates a source file from a file path and text.
     * Adds it to the cache.
     * @param filePath - File path for the source file.
     * @param sourceText - Text to create the source file with.
     * @param options - Options.
     * @throws InvalidOperationError if the file exists.
     */
    createSourceFileFromText(filePath, sourceText, options) {
        filePath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(filePath);
        if (options.overwrite === true)
            return this.createOrOverwriteSourceFileFromText(filePath, sourceText, options);
        this.throwIfFileExists(filePath);
        return this.createSourceFileFromTextInternal(filePath, sourceText, options);
    }
    /**
     * Throws an error if the file exists in the cache or file system.
     * @param filePath - File path.
     * @param prefixMessage - Message to attach on as a prefix.
     */
    throwIfFileExists(filePath, prefixMessage) {
        if (!this.containsSourceFileAtPath(filePath) && !this.context.fileSystemWrapper.fileExistsSync(filePath))
            return;
        prefixMessage = prefixMessage == null ? "" : prefixMessage + " ";
        throw new errors.InvalidOperationError(`${prefixMessage}A source file already exists at the provided file path: ${filePath}`);
    }
    createOrOverwriteSourceFileFromText(filePath, sourceText, options) {
        filePath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(filePath);
        const existingSourceFile = this.addOrGetSourceFileFromFilePath(filePath, options);
        if (existingSourceFile != null) {
            existingSourceFile.getChildren().forEach(c => c.forget());
            this.replaceCompilerNode(existingSourceFile, this.createCompilerSourceFileFromText(filePath, sourceText, options.scriptKind));
            return existingSourceFile;
        }
        return this.createSourceFileFromTextInternal(filePath, sourceText, options);
    }
    /**
     * Gets the source file from the cache by a file path.
     * @param filePath - File path.
     */
    getSourceFileFromCacheFromFilePath(filePath) {
        filePath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(filePath);
        return this.sourceFileCacheByFilePath.get(filePath);
    }
    /**
     * Gets a source file from a file path. Will use the file path cache if the file exists.
     * @param filePath - File path to get the file from.
     */
    addOrGetSourceFileFromFilePath(filePath, options) {
        filePath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(filePath);
        let sourceFile = this.sourceFileCacheByFilePath.get(filePath);
        if (sourceFile == null && this.context.fileSystemWrapper.fileExistsSync(filePath)) {
            this.context.logger.log(`Loading file: ${filePath}`);
            sourceFile = this.createSourceFileFromTextInternal(filePath, this.context.fileSystemWrapper.readFileSync(filePath, this.context.getEncoding()), options);
            sourceFile._setIsSaved(true); // source files loaded from the disk are saved to start with
        }
        if (sourceFile != null && options.markInProject)
            sourceFile._markAsInProject();
        return sourceFile;
    }
    /**
     * Gets if the internal cache contains a source file at a specific file path.
     * @param filePath - File path to check.
     */
    containsSourceFileAtPath(filePath) {
        const absoluteFilePath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(filePath);
        return this.sourceFileCacheByFilePath.has(absoluteFilePath);
    }
    /**
     * Gets if the internal cache contains a source file with the specified directory path.
     * @param dirPath - Directory path to check.
     */
    containsDirectoryAtPath(dirPath) {
        const normalizedDirPath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(dirPath);
        return this.directoryCache.has(normalizedDirPath);
    }
    /**
     * Gets the source file for a node.
     * @param compilerNode - Compiler node to get the source file of.
     */
    getSourceFileForNode(compilerNode) {
        let currentNode = compilerNode;
        while (currentNode.kind !== typescript_1.SyntaxKind.SourceFile) {
            if (currentNode.parent == null)
                return undefined;
            currentNode = currentNode.parent;
        }
        return this.getSourceFile(currentNode, { markInProject: false });
    }
    /**
     * Gets if the factory contains the compiler node in its internal cache.
     * @param compilerNode - Compiler node.
     */
    hasCompilerNode(compilerNode) {
        return this.nodeCache.has(compilerNode);
    }
    /**
     * Gets an existing node from the cache.
     * @param compilerNode - Compiler node.
     */
    getExistingCompilerNode(compilerNode) {
        return this.nodeCache.get(compilerNode);
    }
    /**
     * Gets a wrapped compiler type based on the node's kind.
     * @param node - Node to get the wrapped object from.
     */
    getNodeFromCompilerNode(compilerNode, sourceFile) {
        if (compilerNode.kind === typescript_1.SyntaxKind.SourceFile)
            return this.getSourceFile(compilerNode, { markInProject: false });
        return this.nodeCache.getOrCreate(compilerNode, () => {
            const node = createNode.call(this);
            initializeNode.call(this, node);
            return node;
        });
        function createNode() {
            // todo: improve kind to wrapper mappings to handle this scenario
            if (isCommentNode(compilerNode)) {
                if (utils_1.CommentNodeParser.isCommentStatement(compilerNode))
                    return new compiler_1.CommentStatement(this.context, compilerNode, sourceFile);
                if (utils_1.CommentNodeParser.isCommentClassElement(compilerNode))
                    return new compiler_1.CommentClassElement(this.context, compilerNode, sourceFile);
                if (utils_1.CommentNodeParser.isCommentTypeElement(compilerNode))
                    return new compiler_1.CommentTypeElement(this.context, compilerNode, sourceFile);
                if (utils_1.CommentNodeParser.isCommentObjectLiteralElement(compilerNode))
                    return new compiler_1.CommentObjectLiteralElement(this.context, compilerNode, sourceFile);
                if (utils_1.CommentNodeParser.isCommentEnumMember(compilerNode))
                    return new compiler_1.CommentEnumMember(this.context, compilerNode, sourceFile);
                return errors.throwNotImplementedForNeverValueError(compilerNode);
            }
            const ctor = kindToWrapperMappings_1.kindToWrapperMappings[compilerNode.kind] || compiler_1.Node;
            return new ctor(this.context, compilerNode, sourceFile);
        }
        function isCommentNode(node) {
            return node._commentKind != null;
        }
        function initializeNode(node) {
            // ensure the parent is created and increment its wrapped child count
            if (compilerNode.parent != null) {
                const parentNode = this.getNodeFromCompilerNode(compilerNode.parent, sourceFile);
                parentNode._wrappedChildCount++;
            }
            const parentSyntaxList = node._getParentSyntaxListIfWrapped();
            if (parentSyntaxList != null)
                parentSyntaxList._wrappedChildCount++;
            if (compilerNode.kind === typescript_1.SyntaxKind.SyntaxList) {
                let count = 0;
                for (const _ of node._getChildrenInCacheIterator())
                    count++;
                node._wrappedChildCount = count;
            }
        }
    }
    createSourceFileFromTextInternal(filePath, text, options) {
        const hasBom = utils_2.StringUtils.hasBom(text);
        if (hasBom)
            text = utils_2.StringUtils.stripBom(text);
        const sourceFile = this.getSourceFile(this.createCompilerSourceFileFromText(filePath, text, options.scriptKind), options);
        if (hasBom)
            sourceFile._hasBom = true;
        return sourceFile;
    }
    createCompilerSourceFileFromText(filePath, text, scriptKind) {
        return this.documentRegistry.createOrUpdateSourceFile(filePath, this.context.compilerOptions.get(), typescript_1.ts.ScriptSnapshot.fromString(text), scriptKind);
    }
    /**
     * Gets a wrapped source file from a compiler source file.
     * @param sourceFile - Compiler source file.
     */
    getSourceFile(compilerSourceFile, options) {
        let wasAdded = false;
        const sourceFile = this.nodeCache.getOrCreate(compilerSourceFile, () => {
            const createdSourceFile = new compiler_1.SourceFile(this.context, compilerSourceFile);
            if (!options.markInProject)
                this.context.inProjectCoordinator.setSourceFileNotInProject(createdSourceFile);
            this.addSourceFileToCache(createdSourceFile);
            wasAdded = true;
            return createdSourceFile;
        });
        if (options.markInProject)
            sourceFile._markAsInProject();
        if (wasAdded)
            this.sourceFileAddedEventContainer.fire(sourceFile);
        return sourceFile;
    }
    addSourceFileToCache(sourceFile) {
        this.sourceFileCacheByFilePath.set(sourceFile.getFilePath(), sourceFile);
        this.context.fileSystemWrapper.removeFileDelete(sourceFile.getFilePath());
        this.directoryCache.addSourceFile(sourceFile);
    }
    /**
     * Gets a directory from a path.
     * @param dirPath - Directory path.
     */
    getDirectoryFromPath(dirPath, options) {
        dirPath = this.context.fileSystemWrapper.getStandardizedAbsolutePath(dirPath);
        let directory = this.directoryCache.get(dirPath);
        if (directory == null && this.context.fileSystemWrapper.directoryExistsSync(dirPath))
            directory = this.directoryCache.createOrAddIfExists(dirPath);
        if (directory != null && options.markInProject)
            directory._markAsInProject();
        return directory;
    }
    /**
     * Creates or adds a directory if it doesn't exist.
     * @param dirPath - Directory path.
     */
    createDirectoryOrAddIfExists(dirPath, options) {
        const directory = this.directoryCache.createOrAddIfExists(dirPath);
        if (directory != null && options.markInProject)
            directory._markAsInProject();
        return directory;
    }
    /**
     * Gets a directory.
     * @param dirPath - Directory path.
     */
    getDirectoryFromCache(dirPath) {
        return this.directoryCache.get(dirPath);
    }
    /**
     * Gets a directory from the cache, but only if it's in the cache.
     * @param dirPath - Directory path.
     */
    getDirectoryFromCacheOnlyIfInCache(dirPath) {
        return this.directoryCache.has(dirPath)
            ? this.directoryCache.get(dirPath)
            : undefined;
    }
    /**
     * Gets all the directories iterated by depth.
     */
    getDirectoriesByDepth() {
        return this.directoryCache.getAllByDepth();
    }
    /**
     * Gets the directories without a parent.
     */
    getOrphanDirectories() {
        return this.directoryCache.getOrphans();
    }
    /**
     * Gets a warpped symbol display part form a compiler symbol display part.
     * @param compilerObject - Compiler symbol display part.
     */
    getSymbolDisplayPart(compilerObject) {
        return this.symbolDisplayPartCache.getOrCreate(compilerObject, () => new compiler_1.SymbolDisplayPart(compilerObject));
    }
    /**
     * Gets a wrapped type from a compiler type.
     * @param type - Compiler type.
     */
    getType(type) {
        if ((type.flags & typescript_1.TypeFlags.TypeParameter) === typescript_1.TypeFlags.TypeParameter)
            return this.getTypeParameter(type);
        return this.typeCache.getOrCreate(type, () => new compiler_1.Type(this.context, type));
    }
    /**
     * Gets a wrapped type parameter from a compiler type parameter.
     * @param typeParameter - Compiler type parameter
     */
    getTypeParameter(typeParameter) {
        return this.typeParameterCache.getOrCreate(typeParameter, () => new compiler_1.TypeParameter(this.context, typeParameter));
    }
    /**
     * Gets a wrapped signature from a compiler signature.
     * @param signature - Compiler signature.
     */
    getSignature(signature) {
        return this.signatureCache.getOrCreate(signature, () => new compiler_1.Signature(this.context, signature));
    }
    /**
     * Gets a wrapped symbol from a compiler symbol.
     * @param symbol - Compiler symbol.
     */
    getSymbol(symbol) {
        return this.symbolCache.getOrCreate(symbol, () => new compiler_1.Symbol(this.context, symbol));
    }
    /**
     * Gets a wrapped definition info from a compiler object.
     * @param compilerObject - Compiler definition info.
     */
    getDefinitionInfo(compilerObject) {
        return this.definitionInfoCache.getOrCreate(compilerObject, () => new compiler_1.DefinitionInfo(this.context, compilerObject));
    }
    /**
     * Gets a wrapped document span from a compiler object.
     * @param compilerObject - Compiler document span.
     */
    getDocumentSpan(compilerObject) {
        return this.documentSpanCache.getOrCreate(compilerObject, () => new compiler_1.DocumentSpan(this.context, compilerObject));
    }
    /**
     * Gets a wrapped referenced entry from a compiler object.
     * @param compilerObject - Compiler referenced entry.
     */
    getReferenceEntry(compilerObject) {
        return this.referenceEntryCache.getOrCreate(compilerObject, () => new compiler_1.ReferenceEntry(this.context, compilerObject));
    }
    /**
     * Gets a wrapped referenced symbol from a compiler object.
     * @param compilerObject - Compiler referenced symbol.
     */
    getReferencedSymbol(compilerObject) {
        return this.referencedSymbolCache.getOrCreate(compilerObject, () => new compiler_1.ReferencedSymbol(this.context, compilerObject));
    }
    /**
     * Gets a wrapped referenced symbol definition info from a compiler object.
     * @param compilerObject - Compiler referenced symbol definition info.
     */
    getReferencedSymbolDefinitionInfo(compilerObject) {
        return this.referencedSymbolDefinitionInfoCache.getOrCreate(compilerObject, () => new compiler_1.ReferencedSymbolDefinitionInfo(this.context, compilerObject));
    }
    /**
     * Gets a wrapped diagnostic from a compiler diagnostic.
     * @param diagnostic - Compiler diagnostic.
     */
    getDiagnostic(diagnostic) {
        return this.diagnosticCache.getOrCreate(diagnostic, () => {
            if (diagnostic.start != null)
                return new compiler_1.DiagnosticWithLocation(this.context, diagnostic);
            return new compiler_1.Diagnostic(this.context, diagnostic);
        });
    }
    /**
     * Gets a wrapped diagnostic with location from a compiler diagnostic.
     * @param diagnostic - Compiler diagnostic.
     */
    getDiagnosticWithLocation(diagnostic) {
        return this.diagnosticCache.getOrCreate(diagnostic, () => new compiler_1.DiagnosticWithLocation(this.context, diagnostic));
    }
    /**
     * Gets a wrapped diagnostic message chain from a compiler diagnostic message chain.
     * @param diagnosticMessageChain - Compiler diagnostic message chain.
     */
    getDiagnosticMessageChain(compilerObject) {
        return this.diagnosticMessageChainCache.getOrCreate(compilerObject, () => new compiler_1.DiagnosticMessageChain(compilerObject));
    }
    /**
     * Gets a warpped JS doc tag info from a compiler object.
     * @param jsDocTagInfo - Compiler object.
     */
    getJSDocTagInfo(jsDocTagInfo) {
        return this.jsDocTagInfoCache.getOrCreate(jsDocTagInfo, () => new compiler_1.JSDocTagInfo(jsDocTagInfo));
    }
    /**
     * Replaces a compiler node in the cache.
     * @param oldNode - Old node to remove.
     * @param newNode - New node to use.
     */
    replaceCompilerNode(oldNode, newNode) {
        const nodeToReplace = oldNode instanceof compiler_1.Node ? oldNode.compilerNode : oldNode;
        const node = oldNode instanceof compiler_1.Node ? oldNode : this.nodeCache.get(oldNode);
        if (nodeToReplace.kind === typescript_1.SyntaxKind.SourceFile && nodeToReplace.fileName !== newNode.fileName) {
            const sourceFile = node;
            this.removeCompilerNodeFromCache(nodeToReplace);
            sourceFile._replaceCompilerNodeFromFactory(newNode);
            this.nodeCache.set(newNode, sourceFile);
            this.addSourceFileToCache(sourceFile);
            this.sourceFileAddedEventContainer.fire(sourceFile);
        }
        else {
            this.nodeCache.replaceKey(nodeToReplace, newNode);
            if (node != null)
                node._replaceCompilerNodeFromFactory(newNode);
        }
    }
    /**
     * Removes a node from the cache.
     * @param node - Node to remove.
     */
    removeNodeFromCache(node) {
        this.removeCompilerNodeFromCache(node.compilerNode);
    }
    /**
     * Removes a compiler node from the cache.
     * @param compilerNode - Compiler node to remove.
     */
    removeCompilerNodeFromCache(compilerNode) {
        this.nodeCache.removeByKey(compilerNode);
        if (compilerNode.kind === typescript_1.SyntaxKind.SourceFile) {
            const sourceFile = compilerNode;
            this.directoryCache.removeSourceFile(sourceFile.fileName);
            const wrappedSourceFile = this.sourceFileCacheByFilePath.get(sourceFile.fileName);
            this.sourceFileCacheByFilePath.removeByKey(sourceFile.fileName);
            this.documentRegistry.removeSourceFile(sourceFile.fileName);
            if (wrappedSourceFile != null)
                this.sourceFileRemovedEventContainer.fire(wrappedSourceFile);
        }
    }
    /**
     * Adds the specified directory to the cache.
     * @param directory - Directory
     */
    addDirectoryToCache(directory) {
        this.directoryCache.addDirectory(directory);
    }
    /**
     * Removes the directory from the cache.
     * @param dirPath - Directory path.
     */
    removeDirectoryFromCache(dirPath) {
        this.directoryCache.remove(dirPath);
    }
    /**
     * Forgets the nodes created in the block.
     * @param block - Block of code to run.
     */
    forgetNodesCreatedInBlock(block) {
        // can't use the async keyword here because exceptions that happen when doing this synchronously need to be thrown
        this.nodeCache.setForgetPoint();
        let wasPromise = false;
        try {
            const result = block((...nodes) => {
                for (const node of nodes)
                    this.nodeCache.rememberNode(node);
            });
            if (result != null && typeof result.then === "function") {
                wasPromise = true;
                return result.then(() => this.nodeCache.forgetLastPoint());
            }
        }
        finally {
            if (!wasPromise)
                this.nodeCache.forgetLastPoint();
        }
        return Promise.resolve();
    }
}
exports.CompilerFactory = CompilerFactory;
