// ts-morph/dist/compiler/ast/module/SourceFile.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors = require("../../../errors");
const manipulation_1 = require("../../../manipulation");
const textSeek_1 = require("../../../manipulation/textSeek");
const structures_1 = require("../../../structures");
const typescript_1 = require("../../../typescript");
const utils_1 = require("../../../utils");
const base_1 = require("../base");
const callBaseGetStructure_1 = require("../callBaseGetStructure");
const callBaseSet_1 = require("../callBaseSet");
const common_1 = require("../common");
const statement_1 = require("../statement");
const results_1 = require("./results");
// todo: not sure why I need to explicitly type this in order to get TS to not complain... (TS 2.4.1)
exports.SourceFileBase = base_1.ModuledNode(base_1.TextInsertableNode(statement_1.StatementedNode(common_1.Node)));
class SourceFile extends exports.SourceFileBase {
    /**
     * Initializes a new instance.
     * @private
     * @param context - Project context.
     * @param node - Underlying node.
     */
    constructor(context, node) {
        super(context, node, undefined);
        /** @internal */
        this._isSaved = false;
        /** @internal */
        this._modifiedEventContainer = new utils_1.EventContainer();
        /** @internal */
        this._preModifiedEventContainer = new utils_1.EventContainer();
        /** @internal */
        this._referenceContainer = new utils_1.SourceFileReferenceContainer(this);
        // typescript doesn't allow calling `super` with `this`, so set this after
        this.__sourceFile = this;
        // store this before a modification happens to the file
        const onPreModified = () => {
            this.isFromExternalLibrary(); // memoize
            this._preModifiedEventContainer.unsubscribe(onPreModified);
        };
        this._preModifiedEventContainer.subscribe(onPreModified);
    }
    /**
     * @internal
     *
     * WARNING: This should only be called by the compiler factory!
     */
    _replaceCompilerNodeFromFactory(compilerNode) {
        super._replaceCompilerNodeFromFactory(compilerNode);
        this._context.resetProgram(); // make sure the program has the latest source file
        this._isSaved = false;
        this._modifiedEventContainer.fire(this);
    }
    /** @internal */
    _clearInternals() {
        super._clearInternals();
        clearTextRanges(this._referencedFiles);
        clearTextRanges(this._typeReferenceDirectives);
        clearTextRanges(this._libReferenceDirectives);
        delete this._referencedFiles;
        delete this._typeReferenceDirectives;
        delete this._libReferenceDirectives;
        function clearTextRanges(textRanges) {
            if (textRanges == null)
                return;
            textRanges.forEach(r => r._forget());
        }
    }
    /**
     * Gets the file path.
     */
    getFilePath() {
        return this.compilerNode.fileName;
    }
    /**
     * Gets the file path's base name.
     */
    getBaseName() {
        return utils_1.FileUtils.getBaseName(this.getFilePath());
    }
    /**
     * Gets the file path's base name without the extension.
     */
    getBaseNameWithoutExtension() {
        const baseName = this.getBaseName();
        const extension = this.getExtension();
        return baseName.substring(0, baseName.length - extension.length);
    }
    /**
     * Gets the file path's extension.
     */
    getExtension() {
        return utils_1.FileUtils.getExtension(this.getFilePath());
    }
    /**
     * Gets the directory that the source file is contained in.
     */
    getDirectory() {
        return this._context.compilerFactory.getDirectoryFromCache(this.getDirectoryPath());
    }
    /**
     * Gets the directory path that the source file is contained in.
     */
    getDirectoryPath() {
        return utils_1.FileUtils.getDirPath(this.compilerNode.fileName);
    }
    /**
     * Gets the full text with leading trivia.
     */
    getFullText() {
        // return the string instead of letting Node.getFullText() do a substring to prevent an extra allocation
        return this.compilerNode.text;
    }
    /**
     * Gets the line and column number at the provided position (1-indexed).
     * @param pos - Position in the source file.
     */
    getLineAndColumnAtPos(pos) {
        const fullText = this.getFullText();
        return {
            line: utils_1.StringUtils.getLineNumberAtPos(fullText, pos),
            column: utils_1.StringUtils.getLengthFromLineStartAtPos(fullText, pos) + 1
        };
    }
    /**
     * Gets the character count from the start of the line to the provided position.
     * @param pos - Position.
     */
    getLengthFromLineStartAtPos(pos) {
        return utils_1.StringUtils.getLengthFromLineStartAtPos(this.getFullText(), pos);
    }
    /**
     * Copies this source file to the specified directory.
     *
     * This will modify the module specifiers in the new file, if necessary.
     * @param dirPathOrDirectory Directory path or directory object to copy the file to.
     * @param options Options for copying.
     * @returns The source file the copy was made to.
     */
    copyToDirectory(dirPathOrDirectory, options) {
        const dirPath = typeof dirPathOrDirectory === "string" ? dirPathOrDirectory : dirPathOrDirectory.getPath();
        return this.copy(utils_1.FileUtils.pathJoin(dirPath, this.getBaseName()), options);
    }
    /**
     * Copy this source file to a new file.
     *
     * This will modify the module specifiers in the new file, if necessary.
     * @param filePath - New file path. Can be relative to the original file or an absolute path.
     * @param options - Options for copying.
     */
    copy(filePath, options = {}) {
        const result = this._copyInternal(filePath, options);
        if (result === false)
            return this;
        const copiedSourceFile = result;
        if (copiedSourceFile.getDirectoryPath() !== this.getDirectoryPath())
            copiedSourceFile._updateReferencesForCopyInternal(this._getReferencesForCopyInternal());
        return copiedSourceFile;
    }
    /** @internal */
    _copyInternal(filePath, options = {}) {
        const { overwrite = false } = options;
        const { compilerFactory, fileSystemWrapper } = this._context;
        filePath = fileSystemWrapper.getStandardizedAbsolutePath(filePath, this.getDirectoryPath());
        if (filePath === this.getFilePath())
            return false;
        return getCopiedSourceFile(this);
        function getCopiedSourceFile(currentFile) {
            try {
                return compilerFactory.createSourceFileFromText(filePath, currentFile.getFullText(), { overwrite, markInProject: getShouldBeInProject() });
            }
            catch (err) {
                if (err instanceof errors.InvalidOperationError)
                    throw new errors.InvalidOperationError(`Did you mean to provide the overwrite option? ` + err.message);
                else
                    throw err;
            }
            function getShouldBeInProject() {
                if (currentFile._isInProject())
                    return true;
                const destinationFile = compilerFactory.getSourceFileFromCacheFromFilePath(filePath);
                return destinationFile != null && destinationFile._isInProject();
            }
        }
    }
    /** @internal */
    _getReferencesForCopyInternal() {
        return Array.from(this._referenceContainer.getLiteralsReferencingOtherSourceFilesEntries());
    }
    /** @internal */
    _updateReferencesForCopyInternal(literalReferences) {
        // update the nodes in this list to point to the nodes in this copied source file
        for (const reference of literalReferences)
            reference[0] = this.getChildSyntaxListOrThrow().getDescendantAtStartWithWidth(reference[0].getStart(), reference[0].getWidth());
        // update the string literals in the copied file
        updateStringLiteralReferences(literalReferences);
    }
    /**
     * Copy this source file to a new file and immediately saves it to the file system asynchronously.
     *
     * This will modify the module specifiers in the new file, if necessary.
     * @param filePath - New file path. Can be relative to the original file or an absolute path.
     * @param options - Options for copying.
     */
    copyImmediately(filePath, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const newSourceFile = this.copy(filePath, options);
            yield newSourceFile.save();
            return newSourceFile;
        });
    }
    /**
     * Copy this source file to a new file and immediately saves it to the file system synchronously.
     *
     * This will modify the module specifiers in the new file, if necessary.
     * @param filePath - New file path. Can be relative to the original file or an absolute path.
     * @param options - Options for copying.
     */
    copyImmediatelySync(filePath, options) {
        const newSourceFile = this.copy(filePath, options);
        newSourceFile.saveSync();
        return newSourceFile;
    }
    /**
     * Moves this source file to the specified directory.
     *
     * This will modify the module specifiers in other files that specify this file and the module specifiers in the current file, if necessary.
     * @param dirPathOrDirectory Directory path or directory object to move the file to.
     * @param options Options for moving.
     */
    moveToDirectory(dirPathOrDirectory, options) {
        const dirPath = typeof dirPathOrDirectory === "string" ? dirPathOrDirectory : dirPathOrDirectory.getPath();
        return this.move(utils_1.FileUtils.pathJoin(dirPath, this.getBaseName()), options);
    }
    /**
     * Moves this source file to a new file.
     *
     * This will modify the module specifiers in other files that specify this file and the module specifiers in the current file, if necessary.
     * @param filePath - New file path. Can be relative to the original file or an absolute path.
     * @param options - Options for moving.
     */
    move(filePath, options = {}) {
        const oldDirPath = this.getDirectoryPath();
        const sourceFileReferences = this._getReferencesForMoveInternal();
        const oldFilePath = this.getFilePath();
        if (!this._moveInternal(filePath, options))
            return this;
        this._context.fileSystemWrapper.queueFileDelete(oldFilePath);
        this._updateReferencesForMoveInternal(sourceFileReferences, oldDirPath);
        // ignore any modifications in other source files
        this._context.lazyReferenceCoordinator.clearDirtySourceFiles();
        // need to add the current source file as being dirty because it was removed and added to the cache in the move
        this._context.lazyReferenceCoordinator.addDirtySourceFile(this);
        return this;
    }
    /** @internal */
    _moveInternal(filePath, options = {}) {
        const { overwrite = false } = options;
        filePath = this._context.fileSystemWrapper.getStandardizedAbsolutePath(filePath, this.getDirectoryPath());
        if (filePath === this.getFilePath())
            return false;
        let markAsInProject = false;
        if (overwrite) {
            // remove the past file if it exists
            const existingSourceFile = this._context.compilerFactory.getSourceFileFromCacheFromFilePath(filePath);
            if (existingSourceFile != null) {
                markAsInProject = existingSourceFile._isInProject();
                existingSourceFile.forget();
            }
        }
        else
            this._context.compilerFactory.throwIfFileExists(filePath, "Did you mean to provide the overwrite option?");
        manipulation_1.replaceSourceFileForFilePathMove({
            newFilePath: filePath,
            sourceFile: this
        });
        if (markAsInProject)
            this._markAsInProject();
        if (this._isInProject())
            this.getDirectory()._markAsInProject();
        return true;
    }
    /** @internal */
    _getReferencesForMoveInternal() {
        return {
            literalReferences: Array.from(this._referenceContainer.getLiteralsReferencingOtherSourceFilesEntries()),
            referencingLiterals: Array.from(this._referenceContainer.getReferencingLiteralsInOtherSourceFiles())
        };
    }
    /** @internal */
    _updateReferencesForMoveInternal(sourceFileReferences, oldDirPath) {
        const { literalReferences, referencingLiterals } = sourceFileReferences;
        // update the literals in this file if the directory has changed
        if (oldDirPath !== this.getDirectoryPath())
            updateStringLiteralReferences(literalReferences);
        // update the string literals in other files
        updateStringLiteralReferences(referencingLiterals.map(node => ([node, this])));
    }
    /**
     * Moves this source file to a new file and asynchronously updates the file system immediately.
     *
     * This will modify the module specifiers in other files that specify this file and the module specifiers in the current file, if necessary.
     * @param filePath - New file path. Can be relative to the original file or an absolute path.
     * @param options - Options for moving.
     */
    moveImmediately(filePath, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const oldFilePath = this.getFilePath();
            const newFilePath = this._context.fileSystemWrapper.getStandardizedAbsolutePath(filePath, this.getDirectoryPath());
            this.move(filePath, options);
            if (oldFilePath !== newFilePath) {
                yield this._context.fileSystemWrapper.moveFileImmediately(oldFilePath, newFilePath, this.getFullText());
                this._isSaved = true;
            }
            else
                yield this.save();
            return this;
        });
    }
    /**
     * Moves this source file to a new file and synchronously updates the file system immediately.
     *
     * This will modify the module specifiers in other files that specify this file and the module specifiers in the current file, if necessary.
     * @param filePath - New file path. Can be relative to the original file or an absolute path.
     * @param options - Options for moving.
     */
    moveImmediatelySync(filePath, options) {
        const oldFilePath = this.getFilePath();
        const newFilePath = this._context.fileSystemWrapper.getStandardizedAbsolutePath(filePath, this.getDirectoryPath());
        this.move(filePath, options);
        if (oldFilePath !== newFilePath) {
            this._context.fileSystemWrapper.moveFileImmediatelySync(oldFilePath, newFilePath, this.getFullText());
            this._isSaved = true;
        }
        else
            this.saveSync();
        return this;
    }
    /**
     * Queues a deletion of the file to the file system.
     *
     * The file will be deleted when you call ast.save(). If you wish to immediately delete the file, then use deleteImmediately().
     */
    delete() {
        const filePath = this.getFilePath();
        this.forget();
        this._context.fileSystemWrapper.queueFileDelete(filePath);
    }
    /**
     * Asynchronously deletes the file from the file system.
     */
    deleteImmediately() {
        return __awaiter(this, void 0, void 0, function* () {
            const filePath = this.getFilePath();
            this.forget();
            yield this._context.fileSystemWrapper.deleteFileImmediately(filePath);
        });
    }
    /**
     * Synchronously deletes the file from the file system.
     */
    deleteImmediatelySync() {
        const filePath = this.getFilePath();
        this.forget();
        this._context.fileSystemWrapper.deleteFileImmediatelySync(filePath);
    }
    /**
     * Asynchronously saves this file with any changes.
     */
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._context.fileSystemWrapper.writeFile(this.getFilePath(), this._getTextForSave());
            this._isSaved = true;
        });
    }
    /**
     * Synchronously saves this file with any changes.
     */
    saveSync() {
        this._context.fileSystemWrapper.writeFileSync(this.getFilePath(), this._getTextForSave());
        this._isSaved = true;
    }
    /** @internal */
    _getTextForSave() {
        const text = this.getFullText();
        return this._hasBom ? "\uFEFF" + text : text;
    }
    /**
     * Gets any `/// <reference path="..." />` comments.
     */
    getReferencedFiles() {
        if (this._referencedFiles == null) {
            this._referencedFiles = (this.compilerNode.referencedFiles || [])
                .map(f => new results_1.FileReference(f, this));
        }
        return this._referencedFiles;
    }
    /**
     * Gets any `/// <reference types="..." />` comments.
     */
    getTypeReferenceDirectives() {
        if (this._typeReferenceDirectives == null) {
            this._typeReferenceDirectives = (this.compilerNode.typeReferenceDirectives || [])
                .map(f => new results_1.FileReference(f, this));
        }
        return this._typeReferenceDirectives;
    }
    /**
     * Gets any `/// <reference lib="..." />` comments.
     */
    getLibReferenceDirectives() {
        if (this._libReferenceDirectives == null) {
            this._libReferenceDirectives = (this.compilerNode.libReferenceDirectives || [])
                .map(f => new results_1.FileReference(f, this));
        }
        return this._libReferenceDirectives;
    }
    /**
     * Get any source files that reference this source file.
     */
    getReferencingSourceFiles() {
        return Array.from(this._referenceContainer.getDependentSourceFiles());
    }
    /**
     * Gets the import and exports in other source files that reference this source file.
     */
    getReferencingNodesInOtherSourceFiles() {
        return Array.from(this._referenceContainer.getReferencingNodesInOtherSourceFiles());
    }
    /**
     * Gets the string literals in other source files that reference this source file.
     */
    getReferencingLiteralsInOtherSourceFiles() {
        return Array.from(this._referenceContainer.getReferencingLiteralsInOtherSourceFiles());
    }
    /**
     * Gets all the descendant string literals that reference a source file.
     */
    getImportStringLiterals() {
        this._ensureBound();
        const literals = (this.compilerNode.imports || []);
        return literals.filter(l => (l.flags & typescript_1.ts.NodeFlags.Synthesized) === 0).map(l => this._getNodeFromCompilerNode(l));
    }
    /**
     * Gets the script target of the source file.
     */
    getLanguageVersion() {
        return this.compilerNode.languageVersion;
    }
    /**
     * Gets the language variant of the source file.
     */
    getLanguageVariant() {
        return this.compilerNode.languageVariant;
    }
    /**
     * Gets the script kind of the source file.
     */
    getScriptKind() {
        // todo: open issue on typescript repo about making this not internal?
        // otherwise, store a collection of what each source file should be.
        return this.compilerNode.scriptKind;
    }
    /**
     * Gets if this is a declaration file.
     */
    isDeclarationFile() {
        return this.compilerNode.isDeclarationFile;
    }
    /**
     * Gets if the source file was discovered while loading an external library.
     */
    isFromExternalLibrary() {
        // This needs to be memoized and stored before modification because the TypeScript
        // compiler does the following code:
        //
        // function isSourceFileFromExternalLibrary(file: SourceFile): boolean {
        //    return !!sourceFilesFoundSearchingNodeModules.get(file.path);
        // }
        //
        // So the compiler node will become out of date after a manipulation occurs and
        // this will return false.
        // do not create the program if not created before... if the program is
        // not created then we know this source file wasn't discovered by the program
        if (!this._context.program._isCompilerProgramCreated())
            return false;
        const compilerProgram = this._context.program.compilerObject;
        return compilerProgram.isSourceFileFromExternalLibrary(this.compilerNode);
    }
    /**
     * Gets if the source file is a descendant of a node_modules directory.
     */
    isInNodeModules() {
        return this.getFilePath().indexOf("/node_modules/") >= 0;
    }
    /**
     * Gets if this source file has been saved or if the latest changes have been saved.
     */
    isSaved() {
        return this._isSaved;
    }
    /**
     * Sets if this source file has been saved.
     * @internal
     */
    _setIsSaved(value) {
        this._isSaved = value;
    }
    /**
     * Gets the pre-emit diagnostics of the specified source file.
     */
    getPreEmitDiagnostics() {
        return this._context.getPreEmitDiagnostics(this);
    }
    unindent(positionRangeOrPos, times = 1) {
        return this.indent(positionRangeOrPos, times * -1);
    }
    indent(positionRangeOrPos, times = 1) {
        if (times === 0)
            return this;
        const sourceFileText = this.getFullText();
        const positionRange = typeof positionRangeOrPos === "number" ? [positionRangeOrPos, positionRangeOrPos] : positionRangeOrPos;
        errors.throwIfRangeOutOfRange(positionRange, [0, sourceFileText.length], "positionRange");
        const startLinePos = textSeek_1.getPreviousMatchingPos(sourceFileText, positionRange[0], char => char === "\n");
        const endLinePos = textSeek_1.getNextMatchingPos(sourceFileText, positionRange[1], char => char === "\r" || char === "\n");
        const correctedText = utils_1.StringUtils.indent(sourceFileText.substring(startLinePos, endLinePos), times, {
            indentText: this._context.manipulationSettings.getIndentationText(),
            indentSizeInSpaces: this._context.manipulationSettings._getIndentSizeInSpaces(),
            isInStringAtPos: pos => this.isInStringAtPos(pos + startLinePos)
        });
        manipulation_1.replaceSourceFileTextForFormatting({
            sourceFile: this,
            newText: sourceFileText.substring(0, startLinePos) + correctedText + sourceFileText.substring(endLinePos)
        });
        return this;
    }
    /**
     * Asynchronously emits the source file as a JavaScript file.
     */
    emit(options) {
        return this._context.program.emit(Object.assign({ targetSourceFile: this }, options));
    }
    /**
     * Synchronously emits the source file as a JavaScript file.
     */
    emitSync(options) {
        return this._context.program.emitSync(Object.assign({ targetSourceFile: this }, options));
    }
    /**
     * Gets the emit output of this source file.
     * @param options - Emit options.
     */
    getEmitOutput(options = {}) {
        return this._context.languageService.getEmitOutput(this, options.emitOnlyDtsFiles || false);
    }
    /**
     * Formats the source file text using the internal TypeScript formatting API.
     * @param settings - Format code settings.
     */
    formatText(settings = {}) {
        manipulation_1.replaceSourceFileTextForFormatting({
            sourceFile: this,
            newText: this._context.languageService.getFormattedDocumentText(this.getFilePath(), settings)
        });
    }
    /**
     * Refresh the source file from the file system.
     *
     * WARNING: When updating from the file system, this will "forget" any previously navigated nodes.
     * @returns What action ended up taking place.
     */
    refreshFromFileSystem() {
        return __awaiter(this, void 0, void 0, function* () {
            const fileReadResult = yield this._context.fileSystemWrapper.readFileOrNotExists(this.getFilePath(), this._context.getEncoding());
            return this._refreshFromFileSystemInternal(fileReadResult);
        });
    }
    /**
     * Synchronously refreshes the source file from the file system.
     *
     * WARNING: When updating from the file system, this will "forget" any previously navigated nodes.
     * @returns What action ended up taking place.
     */
    refreshFromFileSystemSync() {
        const fileReadResult = this._context.fileSystemWrapper.readFileOrNotExistsSync(this.getFilePath(), this._context.getEncoding());
        return this._refreshFromFileSystemInternal(fileReadResult);
    }
    getRelativePathTo(sourceFileOrDir) {
        return this.getDirectory().getRelativePathTo(sourceFileOrDir);
    }
    getRelativePathAsModuleSpecifierTo(sourceFileOrDir) {
        return this.getDirectory().getRelativePathAsModuleSpecifierTo(sourceFileOrDir);
    }
    /**
     * Subscribe to when the source file is modified.
     * @param subscription - Subscription.
     * @param subscribe - Optional and defaults to true. Use an explicit false to unsubscribe.
     */
    onModified(subscription, subscribe = true) {
        if (subscribe)
            this._modifiedEventContainer.subscribe(subscription);
        else
            this._modifiedEventContainer.unsubscribe(subscription);
        return this;
    }
    /**
     * Do an action the next time the source file is modified.
     * @param action - Action to run.
     * @internal
     */
    _doActionPreNextModification(action) {
        const wrappedSubscription = () => {
            action();
            this._preModifiedEventContainer.unsubscribe(wrappedSubscription);
        };
        this._preModifiedEventContainer.subscribe(wrappedSubscription);
        return this;
    }
    /** @internal */
    _firePreModified() {
        this._preModifiedEventContainer.fire(this);
    }
    /**
     * Organizes the imports in the file.
     *
     * WARNING! This will forget all the nodes in the file! It's best to do this after you're all done with the file.
     * @param formatSettings - Format code settings.
     * @param userPreferences - User preferences for refactoring.
     */
    organizeImports(formatSettings = {}, userPreferences = {}) {
        this._context.languageService.organizeImports(this, formatSettings, userPreferences).forEach(fileTextChanges => fileTextChanges.applyChanges());
        return this;
    }
    /**
     * Removes all unused declarations like interfaces, classes, enums, functions, variables, parameters,
     * methods, properties, imports, etc. from this file.
     *
     * Tip: For optimal results, sometimes this method needs to be called more than once. There could be nodes
     * that are only referenced in unused declarations and in this case, another call will also remove them.
     *
     * WARNING! This will forget all the nodes in the file! It's best to do this after you're all done with the file.
     * @param formatSettings - Format code settings.
     * @param userPreferences - User preferences for refactoring.
     */
    fixUnusedIdentifiers(formatSettings = {}, userPreferences = {}) {
        this._context.languageService.getCombinedCodeFix(this, "unusedIdentifier_delete", formatSettings, userPreferences).applyChanges();
        return this;
    }
    /**
     * Code fix to add import declarations for identifiers that are referenced, but not imported in the source file.
     * @param formatSettings - Format code settings.
     * @param userPreferences - User preferences for refactoring.
     */
    fixMissingImports(formatSettings = {}, userPreferences = {}) {
        const combinedCodeFix = this._context.languageService.getCombinedCodeFix(this, "fixMissingImport", formatSettings, userPreferences);
        const sourceFile = this;
        for (const fileTextChanges of combinedCodeFix.getChanges()) {
            const changes = fileTextChanges.getTextChanges();
            removeUnnecessaryDoubleBlankLines(changes);
            applyTextChanges(changes);
        }
        return this;
        function removeUnnecessaryDoubleBlankLines(changes) {
            changes.sort((a, b) => a.getSpan().getStart() - b.getSpan().getStart());
            // when a file has no imports, it will add a double newline to every change
            // so remove them except for the last change
            for (let i = 0; i < changes.length - 1; i++) { // skip last change
                const { compilerObject } = changes[i];
                compilerObject.newText = compilerObject.newText.replace(/(\r?)\n\r?\n$/, "$1\n");
            }
        }
        function applyTextChanges(changes) {
            // group all the changes by their start position and insert them into the file
            const groups = utils_1.ArrayUtils.groupBy(changes, change => change.getSpan().getStart());
            let addedLength = 0;
            for (const group of groups) {
                // these should all be import declarations so it should be safe
                const insertPos = group[0].getSpan().getStart() + addedLength;
                const newText = group.map(item => item.getNewText()).join("");
                manipulation_1.insertIntoTextRange({
                    sourceFile,
                    insertPos,
                    newText
                });
                addedLength += newText.length;
            }
        }
    }
    /**
     * Applies the text changes to the source file.
     *
     * WARNING! This will forget all the nodes in the file! It's best to do this after you're all done with the file.
     * @param textChanges - Text changes.
     */
    applyTextChanges(textChanges) {
        // do nothing if no changes
        if (textChanges.length === 0)
            return this;
        this.forgetDescendants();
        manipulation_1.replaceNodeText({
            sourceFile: this._sourceFile,
            start: 0,
            replacingLength: this.getFullWidth(),
            newText: manipulation_1.getTextFromFormattingEdits(this, textChanges)
        });
        return this;
    }
    /**
     * Sets the node from a structure.
     * @param structure - Structure to set the node with.
     */
    set(structure) {
        callBaseSet_1.callBaseSet(exports.SourceFileBase.prototype, this, structure);
        return this;
    }
    /**
     * Gets the structure equivalent to this node.
     */
    getStructure() {
        return callBaseGetStructure_1.callBaseGetStructure(exports.SourceFileBase.prototype, this, {
            kind: structures_1.StructureKind.SourceFile
        });
    }
    _refreshFromFileSystemInternal(fileReadResult) {
        if (fileReadResult === false) {
            this.forget();
            return results_1.FileSystemRefreshResult.Deleted;
        }
        const fileText = fileReadResult;
        if (fileText === this.getFullText())
            return results_1.FileSystemRefreshResult.NoChange;
        this.replaceText([0, this.getEnd()], fileText);
        this._setIsSaved(true); // saved when loaded from file system
        return results_1.FileSystemRefreshResult.Updated;
    }
    /** @internal */
    _isInProject() {
        return this._context.inProjectCoordinator.isSourceFileInProject(this);
    }
    /** @internal */
    _markAsInProject() {
        this._context.inProjectCoordinator.markSourceFileAsInProject(this);
    }
}
__decorate([
    utils_1.Memoize
], SourceFile.prototype, "isFromExternalLibrary", null);
exports.SourceFile = SourceFile;
function updateStringLiteralReferences(nodeReferences) {
    for (const [stringLiteral, sourceFile] of nodeReferences) {
        if (utils_1.ModuleUtils.isModuleSpecifierRelative(stringLiteral.getLiteralText()))
            stringLiteral.setLiteralValue(stringLiteral._sourceFile.getRelativePathAsModuleSpecifierTo(sourceFile));
    }
}
