// ts-morph/dist/main.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference path="typings/index.d.ts" />
__export(require("./codeBlockWriter"));
__export(require("./compiler"));
__export(require("./errors/classes"));
var fileSystem_1 = require("./fileSystem");
exports.Directory = fileSystem_1.Directory;
exports.DirectoryEmitResult = fileSystem_1.DirectoryEmitResult;
__export(require("./options"));
var Project_1 = require("./Project");
exports.Project = Project_1.Project;
__export(require("./structures"));
var createWrappedNode_1 = require("./utils/compiler/createWrappedNode");
exports.createWrappedNode = createWrappedNode_1.createWrappedNode;
var printNode_1 = require("./utils/compiler/printNode");
exports.printNode = printNode_1.printNode;
var getCompilerOptionsFromTsConfig_1 = require("./utils/tsconfig/getCompilerOptionsFromTsConfig");
exports.getCompilerOptionsFromTsConfig = getCompilerOptionsFromTsConfig_1.getCompilerOptionsFromTsConfig;
var utils_1 = require("./utils");
exports.TypeGuards = utils_1.TypeGuards;
const Writers_1 = require("./structurePrinters/Writers");
exports.Writers = Writers_1.Writers;
/** @deprecated Use `Writers`. */
const WriterFunctions = Writers_1.Writers;
exports.WriterFunctions = WriterFunctions;
__export(require("./typescript/public"));
